### PR TITLE
[KYUUBI #3487] Provide Hive JDBC Dialect support for Spark/PySpark to connect Kyuubi via JDBC Source

### LINKED ISSUE #3487: [FEATURE] Provide Hive JDBC Dialect support for Spark/PySpark to connect Kyuubi via JDBC Source
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

Currently, it is not able to read data correctly from Kyuubi as JDBC Dataframe with either KyuubiHiveDriver or HiveDriver.

As Hive JDBC Dialect is not provided in Spark, Spark will force JdbcRDD to use default JDBC dialect which is quoting column name wrapping in column name to 'col_name' instead of Hive style \`col_name\`. It causes two major unexpected situations:

1. if all the columns of table are in string type, all row data is identical to their column name.
2. if any column has number type, it will failed to convert value to number type with wrong data (the column name).

image

image

### Motivation

_No response_

### Describe the solution

Introduce a spark plugin, 
1. providing Hive JDBC Dialet for jdbc url starting with `jdbc:kyuubi://` or `jdbc:hive2://` , with quoting column names in Hive style as \`col_name\`
2. registing Hive JDBC Dialet  by  executing `JdbcDialects.registerDialect` 

### Additional context

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by bowenliang123**
…and register to JdbcDialects

### _Why are the changes needed?_

close #3487 .

1. add kyuubi-extension-spark-client_2.12 module, and introduce KyuubiSparkClientExtension
2. implement HiveDialect and register to JdbcDialects

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by bowenliang123** [comment]
cc @yaooqinn @pan3793

May I ask for some suggestions here?

1. whether appropriate to introduce a new extention plugin here. It's supposed for end users of Spark, PySpark and etc, to support them with HiveDialect and auto registration.
2. any better suggestion for this extention name  other than `kyuubi-extension-spark-client_2.12`.

**Comment by pan3793** [comment]
Thanks, @bowenliang123, I think it's fine to introduce a new module, the "client" is not properly IMO how about `kyuubi-extension-spark-jdbc-dialect_2.12`, under `extensions/spark`?

**Comment by bowenliang123** [comment]
OK, I will take this name `kyuubi-extension-spark-jdbc-dialect_2.12`.
The downside of this name is that it is for dialect specific use and limits further possible for extending features and usage range in client side extension.

File: extensions/client/kyuubi-spark-client/src/main/scala/org/apache/kyuubi/plugin/client/dialect/HiveDialect.scala [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.plugin.client.dialect
+
+import java.util.Locale
+
+import org.apache.spark.sql.jdbc.JdbcDialect
+
+object HiveDialect extends JdbcDialect {
```
Comment (pan3793) [inline extensions/client/kyuubi-spark-client/src/main/scala/org/apache/kyuubi/plugin/client/dialect/HiveDialect.scala:24]:
```suggestion
object KyuubiHiveDialect extends JdbcDialect {
```

File: extensions/client/kyuubi-spark-client/src/main/scala/org/apache/kyuubi/plugin/client/KyuubiSparkClientExtension.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+import org.apache.spark.sql.SparkSessionExtensions
+import org.apache.spark.sql.jdbc.JdbcDialects
+
+import org.apache.kyuubi.plugin.client.dialect.HiveDialect
+
+class KyuubiSparkClientExtension extends (SparkSessionExtensions => Unit) {
+  override def apply(extensions: SparkSessionExtensions): Unit = {
+    // register hive jdbc dialect
+    JdbcDialects.registerDialect(HiveDialect)
```
Comment (pan3793) [inline extensions/client/kyuubi-spark-client/src/main/scala/org/apache/kyuubi/plugin/client/KyuubiSparkClientExtension.scala:28]: How about adding a switch? e.g.`spark.sql.jdbc.KyuubiHiveDialect.enabled`

Comment (bowenliang123) [reply extensions/client/kyuubi-spark-client/src/main/scala/org/apache/kyuubi/plugin/client/KyuubiSparkClientExtension.scala:28]: Good, and maybe we should leave it to `true` by default ?

File: extensions/client/kyuubi-spark-client/src/main/scala/org/apache/kyuubi/plugin/client/dialect/HiveDialect.scala (hunk 1 above)
Comment (bowenliang123) [reply extensions/client/kyuubi-spark-client/src/main/scala/org/apache/kyuubi/plugin/client/dialect/HiveDialect.scala:24]: good idea.

File: extensions/client/kyuubi-spark-client/src/main/scala/org/apache/kyuubi/plugin/client/KyuubiSparkClientExtension.scala (hunk 2 above)
Comment (pan3793) [reply extensions/client/kyuubi-spark-client/src/main/scala/org/apache/kyuubi/plugin/client/KyuubiSparkClientExtension.scala:28]: yea

File: extensions/spark/kyuubi-extension-spark-jdbc-dialect/pom.xml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        <artifactId>kyuubi-parent</artifactId>
+        <groupId>org.apache.kyuubi</groupId>
+        <version>1.7.0-SNAPSHOT</version>
+        <relativePath>../../../pom.xml</relativePath>
+
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>kyuubi-extension-spark-jdbc-dialect_2.12</artifactId>
+    <name>Kyuubi Spark Client-side Extension</name>
```
Comment (pan3793) [inline extensions/spark/kyuubi-extension-spark-jdbc-dialect/pom.xml:32]: nit

File: extensions/spark/kyuubi-extension-spark-jdbc-dialect/pom.xml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-sql_${scala.binary.version}</artifactId>
+            <scope>provided</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <outputDirectory>target/scala-2.12/classes</outputDirectory>
+        <testOutputDirectory>target/scala-2.12/test-classes</testOutputDirectory>
```
Comment (pan3793) [inline extensions/spark/kyuubi-extension-spark-jdbc-dialect/pom.xml:46]:
```suggestion
        <outputDirectory>target/scala-${scala.binary.verison}/classes</outputDirectory>
        <testOutputDirectory>target/scala-${scala.binary.verison}/test-classes</testOutputDirectory>
```

**Review by pan3793 — APPROVED** [review]
LGTM, only minor comments, thanks.

File: extensions/spark/kyuubi-extension-spark-jdbc-dialect/pom.xml (hunk 3 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-extension-spark-jdbc-dialect/pom.xml]: changed to `Kyuubi Spark JDBC Dialect plugin`

File: extensions/spark/kyuubi-extension-spark-jdbc-dialect/pom.xml (hunk 4 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-extension-spark-jdbc-dialect/pom.xml]: applied.

**Comment by bowenliang123** [comment]
Doc for plugin has been added. This PR should be ready for review or merge. @pan3793

**Comment by pan3793** [comment]
Thanks, merging to master

**Pull request closed without merging**